File: ruby/m101/hw1-3.rb
# Author: Marcelo Lipienski

require 'mongo'
require 'sinatra'

get '/hw1/:skip_entry' do
  # connect to the db on standard port
  connection = Mongo::Connection.new('localhost', 27017)

  db = connection.db("m101")                  # attach to db
  collection = db.collection("funnynumbers")  #specify the collection

  doc = collection.find({}, sort: 'value', skip: params[:skip_entry].to_i).next
  response = doc['value'].to_i unless doc.nil?
  
  "#{response}" # outputs response to the browser
  
end

File: golang/src/m101/hw1-3.go
package main

import (
	"fmt"
	"strconv"
	"code.google.com/p/goweb/goweb"
	"github.com/garyburd/go-mongo/mongo"
)

func getHw1(number int) string {
	conn, err := mongo.Dial("localhost")
	if err != nil {
		fmt.Println("Error trying to connect to database")
	}

	defer conn.Close()

	db := mongo.Database{conn, "m101", mongo.DefaultLastErrorCmd}

	funnynumbers := db.C("funnynumbers")

	cursor, err := funnynumbers.Find(nil).Limit(1).Skip(number).Sort(mongo.D{{"value", 1}}).Cursor()
	if err != nil {
		fmt.Println("Error trying to read collection")
	}

	defer cursor.Close()

	var value string
	if cursor.HasNext() {
		var m map[string]interface{}

		cursor.Next(&m)

		value = fmt.Sprint(m["value"])
	}

	return value
}

func main() {
	goweb.MapFunc("/hw1/{number}",
		func(c *goweb.Context) {
			number, err := strconv.ParseInt(c.PathParams["number"], 10, 32)
			if err != nil {
				fmt.Fprintln(c.ResponseWriter, "Expecting valid number")
			}
			fmt.Fprintln(c.ResponseWriter, getHw1(int(number)))
		})

	goweb.ListenAndServe("127.0.0.1:8080")
}
